Iframe: try role=application

ellatrix opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Let's try role=application on the iframe. See https://github.com/WordPress/gutenberg/issues/46260#issuecomment-1365307498.

This PR also makes an adjustment to the `RichText` component. The component will now use `role="textbox"` by default with the option to override if necessary. However, `role="group"` is no longer a valid property here to avoid it being passed down from `useBlockProps` hook directly on `RichText`. Code comment added to explain.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

To fine tune accessibility.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

ellatrix committed:
Iframe: try role=application

alexstine reviewed (commented):
@ellatrix Thanks for trying this out. The initial results are promising.

The biggest problems I notice come from content that is rendered outside of the portal/iFrame. We could reduce a lot of verbosity if we were able to keep more inside the iFrame.

Is there any conceivable way to move the block toolbar inside the iFrame? It should really be attached to each block, the currently selected block, but I know discussions before indicated this is not really possible.

This is for sure an improvement as the screen reader only announces the application now vs. the iFrame, content body, and document of the block. We could make it even better if we could figure out how to keep the most used tools inside the iFrame that way screen reader does not announce every time you leave/enter again. The sidebar has to stay outside of the iFrame for obvious reasons but I wonder if the block toolbar could somehow join.

I would also like feedback from @afercia about how this behaves on Mac. It should be a better experience but I want to make sure of that.

Things are still crazy with the navigation block and I'm not sure why. The `role="application"` should force Windows screen readers into focus mode at all times unless the user changes it. However, after inserting the navigation block, focus gets placed on the block appender causing a mode change. I'm not sure if this is some bug I found in NVDA or maybe if this can be caused when focus is placed directly on a button, even inside of the application role. Either way, the PR does not fix this bug and we're going to need to explore this in another PR/issue. I was just silently hoping the insane navigation block might see some improvements from this.

Moving forward:
1. Wait on @afercia review for Mac.
2. See if there is any good way to move the block toolbar into the iFrame DOM. Maybe try to position it above the title field, as it is now? Might have to play around with the toolbar tabindex to ensure focus does not enter it when tabbing into the application.

Removing the title also made things a lot quieter. I prefer it this way and hopefully it can stay this way. I think we are okay since the application is the first item in the Editor content region.

ellatrix commented:
Rendering the block toolbar inside iframe defeats some of its purpose: a boundary between content and editor UI. We don't render the toolbar in the iframe for the same reason that we don't render the inspector in the iframe, it's not content but an interface for controlling the content that is also styled differently. The iframe should be seen as ideally identical to the front end.

Maybe the editor as a whole should be wrapped into role application?

alexstine commented:
@ellatrix 

> Maybe the editor as a whole should be wrapped into role application?

There has always been a lot of thought about this but also a lot of hesitation. When the whole editor is wrapped in this role, we'd have to manage every last keyboard event perfectly. I think at this point, Slack only wraps the message list and threads list but nowhere else. As I mentioned, adding it to the iFrame is certainly a good start but I just wanted to explore if it was possible to add the block toolbar because that would eliminate a bit more verbosity.

One thing worth noting, the toolbar is actually inside the application role for the classic editor. Any idea how that is being done?

Thanks.

ellatrix commented:
> One thing worth noting, the toolbar is actually inside the application role for the classic editor. Any idea how that is being done?

TinyMCE wraps its whole editor in role=application, which includes the toolbar and the TinyMCE iframe. That's why I suggested wrapping everything in it. Maybe an alternative is wrapping the iframe but additionally the toolbar as well?

alexstine commented:
@ellatrix If it's easy enough to try that, I think we should.

afercia commented:
It is important to understand that role=application is risky, as it 'nullifies' any semantics perceived by screen readers, or at least by some of them. As such, many native screen readers features don't work any longer when inside a role=application. For example:
- no landmarks listing and navigation, because no landmark is perceived
- no forms elements listing navigation, because no forms elements are perceived
- no links listing and navigation, because no links are perceived
- etc.

Basically, most native screen readers navigational tools will not work any longer.

I think role=application is OK to use only in very limited cases, only as last resort, and only on _small_ parts of the UI, when there is the need to force screen readers to switch to a sort of 'focus mode' rather than stayin in browse mode (note the terminology varies dependindg on the screen reader).

Such usage has been done for example for the List View, see https://github.com/WordPress/gutenberg/pull/44291. In that case, role=application may help and it's limited to only this specific part of the UI. The List View is a complex widget and it's more usable when screen readers stop intercepting key strokes.

However, setting role=application on the entire Editor would have, in my opinion, a terrible impact as most native screen readers features would not work any longer. I'd strongly recommend against this proposal.

alexstine commented:
@afercia I tend to agree and that's why this started out as just wrapping the editor content area. I still think it would be wise to somehow place the block toolbar in the editor content region to reduce verbosity moving out/in of the iFrame. The big problem with the editor overall is most of the interactions you described with screen readers don't really work in Gutenberg to begin with, hints the idea that maybe we should be managing the entire experience. For example, switch to browse mode in NVDA after inserting some blocks and tell me if you can tell them apart. You can't because it is largely a bunch of divs with labels, no semantic elements.

I think efforts should continue here to move the block toolbar and we'll wait on trying `role="application"` globally.

Thanks.

ellatrix commented:
Yes, this is only using it on the iframe, which mean it's only for editor content.

> * no landmarks listing and navigation, because no landmark is perceived
> * no forms elements listing navigation, because no forms elements are perceived
> * no links listing and navigation, because no links are perceived
> * etc.

Tbh, it makes sense to me that all these things don't work in the editor content. This is content to be edited, not elements that are part of the interface.

Regarding the toolbar, I still don't quite understand the benefit of role=application. Is that verbosity when switching useless? It seems important to know when it's switching from content to UI? I can try wrapping the toolbar separately. Would that work? Or does it have to be part of the same role=application parent element?

ellatrix commented:
Is it worth adding role=application first to the iframe, then see if we expand it to toolbar? Or is it not so useful for just the iframe?

alexstine commented:
@ellatrix 

> Tbh, it makes sense to me that all these things don't work in the editor content. This is content to be edited, not elements that are part of the interface.

Not sure how familiar you are with Microsoft Word, but there is actually an option with screen readers to have the best of both worlds. Screen readers can either operate in edit mode where the whole document is a canvas or they can operate in browse/virtual mode where you can jump to lists, tables, links, etc. The problem with Gutenberg is it gives sighted people the idea of a wide open canvas but it isn't that for screen reader users. It is a bunch of blocks that all have different interaction models. For example, its not like Gutenberg is one huge text field that you can insert different elements into, it is a bunch of parts that make a final picture. A practical example might be inserting a table block. Visually, it probably looks like a table would on the front-end but it does not have actual table markup in the editor. This makes it fairly difficult for screen reader users to know at any given time what row or column they are editing in and it takes away the ability to use virtual navigation techniques to move around in the block. On the front-end, you can press "T" to find the first table and then move around the columns and rows with table navigation commands.

> Regarding the toolbar, I still don't quite understand the benefit of role=application. Is that verbosity when switching useless? It seems important to know when it's switching from content to UI? I can try wrapping the toolbar separately. Would that work? Or does it have to be part of the same role=application parent element?

It becomes annoying in this situation because every time you enter the toolbar and then return to the iFrame, the screen reader announces the surrounding context of re-entering the iFrame. If I want to bold text in a paragraph block, I don't need to hear that I am entering the iFrame again as I never expected to leave it in the first place. I expect to hear I am in a formatting bar but it is reasonable to expect that if I am editing a block, I will return to said block. Its a balance between not enough information and too much information.

It would be great if the toolbar can appear in the DOM, just inside the iFrame so that extra verbosity goes away on return to editing content.

Thanks.

alexstine commented:
Here is a quick example that might help highlight my point. As you can see, this adds very clear structure to each block and because the table block in my example is coded like an actual table, users could use screen reader virtual mode to navigate to the different fields. Instead of just rendering divs with the `contenteditable` attribute. This is real and valid HTML.

```html
<html>
	<head>
		<title>Test</title>
	</head>
	<body>
		<main>
			<h1>Gutenberg</h1>
			<h2>Post Title</h2>
			<textarea name="post_title" aria-label="Post Title"></textarea>
			<h2>Paragraph</h2>
			<textarea name="paragraph_block_1" aria-label="Paragraph Block">Some paragraph text</textarea>
			<h2>Table Block</h2>
			<table>
				<tr>
					<th><input type="text" name="table_block_1_row_1_header_1" value="Table Header 1" /></th>
					<th><input type="text" name="table_block_1_row_1_header_2" value="Table Header 2" /></th>
				</tr>
				<tr>
					<td><input type="text" name="table_block_1_row_2_cell_1" value="Table Cell 1" /></td>
					<td><input type="text" name="table_block_1_row_2_cell_2" value="Table Cell 2" /></td>
				</tr>
			</table>
		</main>
	</body>
</html>
```

alexstine commented:
The other problem we must solve for at some point is figuring out how to communicate block previews to screen readers. Currently, the only way to make `! isSelected` is to not place focus on the block/deselect it. For obvious reasons, this doesn't work for keyboard-only users. It might be worth exploring some type of preview mode in a follow-up.

alexstine commented:
> Things are still crazy with the navigation block and I'm not sure why. The role="application" should force Windows screen readers into focus mode at all times unless the user changes it. However, after inserting the navigation block, focus gets placed on the block appender causing a mode change. I'm not sure if this is some bug I found in NVDA or maybe if this can be caused when focus is placed directly on a button, even inside of the application role. Either way, the PR does not fix this bug and we're going to need to explore this in another PR/issue. I was just silently hoping the insane navigation block might see some improvements from this.

I tracked this down to the `role="document"` attached to every block. If I remove it, the issue above for Windows screen readers goes away combined with the application role.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/document_role

> Only useful on focusable sections within complex composite widgets or applications, the document role inform assistive technologies to the reading context back to a reading mode: The document role tells assistive technologies with reading or browse modes to use the document mode to read the content contained within this element.

There is nothing to read within the navigation block so this role is actually an anti-pattern. I know @afercia has brought this up specifically before. I don't see much downside to removing it but certain blocks such as table get a little weird due to the announcing of the figure element the block is wrapped in. Part of me wants to remove this role and deal with the fallout vs. keeping this role and all the pain that comes with it. If you insert a paragraph block, it gets the role of textbox after removing the document role. Textbox role is for sure more valid than the generic document role defined as a required value in `useBlockProps`.

I think I'll play around a little with this over the weekend and see what the support looks like across Mac and Windows. I know there has to be a reason for the generic document role but I wonder if said reason is still valid these days.

alexstine committed:
Try removing document role from useBlockProps to avoid Windows screen readers switching modes.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/index.js:
```diff
@@ -183,7 +183,6 @@ export function useBlockProps( props = {}, { __unstableIsHtml } = {} ) {
 		...props,
 		ref: mergedRefs,
 		id: `block-${ clientId }${ htmlSuffix }`,
-		role: 'document',
```
Eliminating this may be actual trouble for a little while until we could work through all blocks and fix any accessibility issues. The problem is, keeping this is hurting us because trying to define a generic role for all the different blocks is not a reasonable solution. For example, this role makes no sense for the paragraph block because the paragraph block is an editable field and this role tells Windows screen readers to go into a mode which is not used for editing text, but navigating the virtual buffer of the page. Removing this also fixes the bug in the navigation block and other dynamic blocks where screen reader switches modes.

ellatrix replied:
It looks like we're always overriding the role to document which seems like a bug. Some blocks like paragraph define role=textbox (from rich text), but useBlockProps is overriding this. Maybe we should only set role=document if no role was defined by the block?

alexstine replied:
@ellatrix I don't think that is possible at this point due to my comments around the navigation block. I think this should likely be managed block by block until we can find a suitable role that could work for any type of content. The document role causes Windows screen readers to switch modes and then the keyboard events no longer work in React so it would be good to avoid this pattern. I think it used the group role before this change and it needed to be modified for a bug in JAWS, another Windows screen reader.

ellatrix replied:
If you're sure about removing role=document from all blocks, I'll update the e2e tests. There's lots of tests depending on this in their selectors.

alexstine replied:
Let's wait on a good accessibility review from more people than just me before you commit to that. I will try to wrangle some testing here.

alexstine commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -259,7 +260,7 @@ function Iframe( {
 				// mode. Also preload the styles to avoid a flash of unstyled
 				// content.
 				src={ src }
-				title={ __( 'Editor canvas' ) }
```
Turns out this is actually an accessibility violation. This isn't a rule I agree with considering the iFrame is wrapped in a `role="region"` with a valid `aria-label`.

@afercia You think this is something we can get away with given the situation?

Having a region labeled as Editor Content followed by the iFrame labeled as Editor Canvas is very verbose and not helpful. The other option, remove the `aria-label` from this region and the title should be good enough since it is the only element?

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/region_role

Regions seem to require accessible labels so really struggling here as to how labeling should:
A. Not be annoying.
B. Not be an accessibility violation from the lack of labeling.

alexstine reviewed (commented):
A couple added thoughts now that I've had real time to test this.

ellatrix commented:
The table in the editor is valid table markup? It's not divs. Here's the markup:

```html
<table>
  <tbody>
    <tr>
      <td role="textbox" aria-multiline="true" aria-label="Body cell text" class="block-editor-rich-text__editable wp-block-table__cell-content rich-text" contenteditable="true">﻿</td>
      <td role="textbox" aria-multiline="true" aria-label="Body cell text" class="block-editor-rich-text__editable wp-block-table__cell-content rich-text" contenteditable="true">﻿</td>
    </tr>
    <tr>
      <td role="textbox" aria-multiline="true" aria-label="Body cell text" class="block-editor-rich-text__editable wp-block-table__cell-content rich-text" contenteditable="true">﻿</td>
      <td role="textbox" aria-multiline="true" aria-label="Body cell text" class="block-editor-rich-text__editable wp-block-table__cell-content rich-text" contenteditable="true">﻿</td>
    </tr>
  </tbody>
</table>
```

> It becomes annoying in this situation because every time you enter the toolbar and then return to the iFrame, the screen reader announces the surrounding context of re-entering the iFrame.

And I guess removing the title attribute doesn't fix this. The next thing to try that I'm hoping works is adding role=application to just the block toolbar. But it is important to understand that you are leaving the block canvas? You don't want to think you're editing the block toolbar? Visually that is understood because it's styled like the admin and it floats around above the canvas, I'm not sure how that would make sense non visually.

> The other problem we must solve for at some point is figuring out how to communicate block previews to screen readers. Currently, the only way to make ! isSelected is to not place focus on the block/deselect it.

I don't understand what you mean by communicating block previews. Could you elaborate?

alexstine commented:
@ellatrix Table might have been a bad example. I believe the table specifically doesn't work due to the fact it is wrapped in a `<figure>` tag. If that was removed, it might work. Probably need to open a tracking issue for block accessibility in general for the back-end.

Block previews. If you deselect the block with the mouse, some blocks can render different content based on the `! isSelected` prop. There is currently no way to access this content because placing keyboard focus on the block will switch the prop to `isSelected`.

Yes, what you mention about the block toolbar does make sense. Also, removing the `title` attribute does eliminate some verbosity considering the iFrame is the only element within the `role="region" aria-label="Editor Content"` wrapper. Having the `title` attribute delivers duplicate information. When you hear announcements about entering Editor Content and then entering Editor Canvas, its essentially the same thing more or less?

CC @joedolson for further input. I could use some assistance thinking through the block toolbar and how to best present it to screen reader users.

We're making progress. Thanks.

ellatrix commented:
> Table might have been a bad example. I believe the table specifically doesn't work due to the fact it is wrapped in a `<figure>` tag. If that was removed, it might work. Probably need to open a tracking issue for block accessibility in general for the back-end.

But how come the table works fine on the front-end, where it is also wrapped in a figure?

ellatrix committed:
Block contextual toolbar: add role=application

ellatrix commented:
@alexstine I rebased this PR and added role=application to the block toolbar's popover. Can't add it to the toolbar directly, I think, because it already has role=toolbar? I have no idea what I'm doing... 🙂

alexstine commented:
@ellatrix I'll test this weekend, no worries.

Not sure why the table block loses its semantic meaning on the back-end and it works fine on the front-end. There are other blocks that have this problem though and its worth looking into why. Probably elsewhere. If you opened up the editor today, inserted a table block, and then used your quick navigation keys to find the table, it would not work.

Sorry for drifting though. I need to keep focus on the current issue so eventually I'll explore this elsewhere.

Thanks.

joedolson commented:
It'll probably need auditing globally; but removing `role=document` seems to be helpful in my testing, as well. Certainly with the paragraph block, it helps by removing some unhelpful verbosity. My first impression is that this is a plausible direction to go, just needs more detail.

Re: the table block. I assume that all of the aria roles are breaking the table semantics. A table cell with `role=textbox` is no longer a table cell. For editing purposes, the roles probably need to be on an inner container so that the table semantics aren't lost. 

I did a quick test on CodePen, and a table where all the cells have `role="textbox"` behaves pretty strangely - it reports as having zero columns and zero rows, regardless of the content contained. Which is pretty much what I expected.

alexstine commented:
Thanks @joedolson. That gave me just enough to figure out the table issue. Opened another PR to take care of it. #54324

I will continue with testing this PR specifically over the weekend.

alexstine commented:
@ellatrix I think its okay to revert changes here.

`packages/block-editor/src/components/block-tools/selected-block-popover.js`

I think my end goal was to have the toolbar inside the `role="application"` and then the iFrame appearing just below that in the application region. This would have helped reduce verbosity but I think given how difficult it will be to introduce the block toolbar in a different DOM position, we can simply exclude it from this PR.

I think the changes to `iFrame` and `useBlockProps` is enough to deal with for the moment.

Thanks.

afercia commented:
Thie Pulls request is trning into a very long discussion that is hard to read with my limited vision. Apoligiesin advance if I'm missing something important.

> For example, switch to browse mode in NVDA after inserting some blocks and tell me if you can tell them apart. You can't because it is largely a bunch of divs with labels, no semantic elements.

> I tracked this down to the role="document" attached to every block. 

@alexstine Since the very beginning of this project we tried to make all the blocks based on RichTexr be perceived as textarea elements. It was our idea that was the most intuitive and exepected semantics and interaction. To do so, the RIchText, which is actually an editable field, usd to use role="textbox" and aria-multiline="true". Tha also made screen readers switch to focus mode, which is expected. I would like to remind you that you wanted to change role="textbox" to role="document" to solve some JAWS bug. Of course, role=document makes these editable divs have no semantics and makes screen readers stay in browse mode. I think that should be reverted to role="textbox". In my experience, trying to solve browsers or assistive technologies bugs by hacking around is often not ideal as it often introduces other problems. 


> Tbh, it makes sense to me that all these things don't work in the editor content. This is content to be edited, not elements that are part of the interface.

@ellatrix i deeply disagree with this point. Nullifying important native screen readers functionalities when inside the content is a no-go for me. This is not a Word process. It is still a web-based application. It does uses form elements, or the ARIA equivalents of them. Pretending to make it behave likw Microsoft Word may work visually, but it will never work semantically and for accessibility.

@alexstine 
Regarding the region labelled `Editor Content` and the iframe labelled `Editor Canvas`. I do realize it is verbose. However, the root problem is not because od this labelling. The root problem is that the editor now users an iframe.

To me, the introduction of the iframe is maybe one of the most wrong decisions ever made in the editor. Users needs should prevaile on developers need. The iframe was only introduced to avoid CSS 'bleeding' between CSS used for the content and the CS used for the interface. While there is some value in that, the iframe is a subpar experience for assistive technology users and shouldn't have been introduced in the first place. Users needs above all.

Technically, the `Editor Content` region actually contains 3 things:
- the notices area
- the slot for the blocks toolbar, when not fixed at the top
- the editor iframe 

This isn't great. As you pointed out, each time users will move from the iframe to the block toolnbar or notice area and back, the will hear the iframe being announced. Sometimes the document title will be announced as well. Actually, the editor is now made of two separete documents. Asrhitecturally, not the best decision I would say. I would totally agree that at least the floating block toolnbar and the content should live in the same document. That would solve a part of the root issue.

Removing the title attribute from the iframe could alleviate the verbosity problem but it would be detected as a violation by any automated accessibility checker. Not opposed to experiment that if it actually helps users.


> added role=application to the block toolbar's popover.

Not sure how that would ever help. The toolbat already has a role=toolbar, which triggers a sort of focus mode for screen readers thusa allowing the expected keyboard interaction. Glad it was removed.

> Not sure why the table block loses its semantic meaning on the back-end and it works fine on the front-end.

As @joedolson pointed out, the unexpected roles on the table cells have likely a deep impact. Worth also reminding that some CSS properties actually nullity the table semantics.

alexstine commented:
@afercia 

> In my experience, trying to solve browsers or assistive technologies bugs by hacking around is often not ideal as it often introduces other problems.

Agreed. I fully own that this was my problem as I made the change. However, I am not sure the role it had before, `role="group"` was much better. No role might be the way to go here.

https://github.com/WordPress/gutenberg/pull/33627/files#diff-2840d9dde2ad8b676f1008acd75d77aca1349194be36689baace2c1f64afcef0L154

Docs: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/group_role

I can kind of see where blocks are groupings of sorts but blocks are not `fieldsets`.

> Most closely related to HTML's <fieldset> element, the group document structure role is used to identify a set of user interface objects which, as compared to region, is not intended to be included in the page's summary or table of contents.

I removed the group role in favor of document role and now I suggest no role. I am not sure when these blocks ever had the textbox role but it was before my change as the diff shows.

I fixed the table block in a follow-up PR: https://github.com/WordPress/gutenberg/pull/54324.

I agree about the regions.

As far as the Microsoft Word editing experience, my comment here explains why Gutenberg is not TinyMCE.

https://github.com/WordPress/gutenberg/pull/54324#issuecomment-1719603791

The entire canvas is content editable where Gutenberg is not.

Re, long PR, I know. Sorry about that. I have a lot of ideas on how to improve the experience without much background from when Gutenberg started. I should have thought better of using this PR as my creative canvas.

Thanks.

ellatrix commented:
> To me, the introduction of the iframe is maybe one of the most wrong decisions ever made in the editor. Users needs should prevaile on developers need. The iframe was only introduced to avoid CSS 'bleeding' between CSS used for the content and the CS used for the interface. While there is some value in that, the iframe is a subpar experience for assistive technology users and shouldn't have been introduced in the first place. Users needs above all.

@afercia It's not only for CSS bleeding, that's just a nice side effect. The main problem this solves and is the ability to preview relative CSS units correctly. Without this iframe these are broken. As you may remember, the classic editor was also iframed. We made the mistake to not take this over from the start.

alexstine committed:
iFrame title adjustment.

alexstine committed:
Revert toolbar changes.

alexstine commented:
- Refreshed the PR for further testing.
- Gave the iFrame maybe a better title for now. The floating block toolbar appears in the Editor content region so I was wrong about the iFrame being the only existing element. Better not remove `title` for now as much as that hurts.
- Still testing the `role="document"` removal. It would be good to get input from a JAWS tester. Before, it used `role="group"` and that caused issues. It now uses `role="document"` and that also caused issues. Now each block can decide which role it uses with no default set. I am open to ideas for default roles excluding group or document. The `RichText` component uses `role="textbox"` by default and that seems to be producing much better results.
- @afercia I finally tracked down the PR where a default `role="group"` was introduced.
https://github.com/WordPress/gutenberg/pull/19213
There is sadly no context here so hard to say what I could have broken when I tried to patch JAWS or what the impact could be to remove the role all together. Sigh.

alexstine commented on packages/block-editor/src/components/block-list/use-block-props/index.js:
```diff
@@ -183,7 +183,6 @@ export function useBlockProps( props = {}, { __unstableIsHtml } = {} ) {
 		...props,
 		ref: mergedRefs,
 		id: `block-${ clientId }${ htmlSuffix }`,
-		role: 'document',
```
Could try to change this back to group role but would need to land this first.

https://github.com/WordPress/gutenberg/pull/47276

If not, JAWS compatibility may break.

alexstine committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into try/iframe-role-application

joedolson commented:
If this needs JAWS testing, I can do a bit of that; I have a JAWS license, at any rate, though I'm far from proficient with it.  Still, I could try and identify possible problems introduced by this change.

alexstine commented:
@joedolson Let me try to refresh this Friday or over the weekend and we can both do some more testing. At some point, we'll need to get into a state where the iFrame is an application and I remove the document role from blocks. That was my sin sadly trying to fix older JAWS versions so I think we'll set it back to group for non-content editable blocks. For content editable blocks, I say we let the rich text role take over which is textbox in most cases but can have additional attributes such as `aria-multiline="true/false"`.

I also agree with @afercia, an issue is probably needed but at this point, there are an endless number of issues that exist from my bad choice, the person before me who made a bad choice, and then of course, the iFrame.

Might be best to close out this PR and draft a new one so I can organize my thoughts. I don't always do the best in writing, that's the dev in me I guess.

alexstine committed:
Refresh.

alexstine committed:
Try using textbox role everywhere by strict comparison of group exclusion.

alexstine commented:
@joedolson Okay, I think this is finally ready for some intense screen reader testing.

Summary:
1. iFrame now has `role="application"`.
2. The `useBlockProps` hook now sets `role="group"` by default.
3. It is now impossible for the `RichText` component to use `role="group"`.
4. The above change was necessary because `RichText` consumes `useBlockProps` hook so the prop will always be whatever we return from that hook by default. Not `role="textbox"`. I decided to let that `RichText` role be customizeable as long as the value doesn't equal group but maybe textbox is solid enough to use everywhere. I want to be cautious making that wide of an assumption at this point.

Please let me know your thoughts. Early testing is looking good with NVDA. Interested to hear if JAWS behaves okay.

Thanks.
